File: cpp/n_queen.cpp
#include <iostream>
using namespace std;

const int N = 4;

bool isSafe(char mat[N][N], int r, int c) {
    for(int i = r; i >= 0; i--) {
        if (mat[i][c] == 'Q') {
            return false;
        }
    }

    for(int i = r, j = c; i >= 0 && j >= 0; i--, j--) {
        if (mat[i][j] == 'Q') {
            return false;
        }
    }

    for(int i = r, j = c; i >= 0 && j < N; i--, j++) {
        if(mat[i][j] == 'Q') {
            return false;
        }
    }
    return true;
}

void nQueen(char mat[N][N], int r) {
    if(r == N) {
        for(int i = 0; i < N; i++) {
            for(int j = 0; j < N; j++) {
                cout<<mat[i][j]<<" ";
            }
            cout<<"\n";
        }
        cout<<"\n";

        return;
    }

    for(int i = 0; i < N; i++) {
        if(isSafe(mat, r, i)) {
            mat[r][i] = 'Q';
            nQueen(mat, r+1);
            mat[r][i] = '-';
        }
    }
}

int main() {
    char mat[N][N];
    for(int i = 0; i < N; i++) {
        for(int j =0; j < N; j++) {
            mat[i][j] = '-';
        }
    }

    nQueen(mat, 0);

    return 0;
} 

File: rust/n_queen.rs
fn main() {
    let n: i32 = 4;
    let mut mat = vec![vec!['-'; n as usize]; n as usize];
    let r: i32 = 0;
    n_queen(&mut mat, r, n);
}

fn is_safe(mat: &mut Vec<Vec<char>>, r: i32, c: i32, n: i32) -> bool {
    let mut i = r;
    while i >= 0 {
        if mat[i as usize][c as usize] == 'Q' {
            return false;
        }
        i -= 1; 
    }

    let mut i = r;
    let mut j = c;
    while i >= 0 && j >= 0 {
        if mat[i as usize][j as usize] == 'Q' {
            return false;
        }
        
        i -= 1;
        j -= 1;
    }

    let mut i = r;
    let mut j = c;
    while i >= 0 && j < n {
        if mat[i as usize][j as usize] == 'Q' {
            return false;
        }

        i -= 1;
        j += 1;
    }

    true
}

fn n_queen(mat: &mut Vec<Vec<char>>, r: i32, n: i32) {
    if r == n {
        println!("{:?}",mat);
        return;
    }

    for i in 0..n {
        if is_safe(mat,r,i, n) {
            mat[r as usize][i as usize] = 'Q';
            n_queen(mat, r+1, n);
            mat[r as usize][i as usize] = '-';
        }
    }
    return;
}